flight/Engine.php

```diff
@@ -364,15 +364,17 @@ public function path(string $dir): void
     /**
      * Processes each routes middleware.
      *
-     * @param array<int, callable> $middleware Middleware attached to the route.
-     * @param array<mixed> $params `$route->params`.
+     * @param Route $route The route to process the middleware for.
      * @param string $event_name If this is the before or after method.
      */
-    protected function processMiddleware(array $middleware, array $params, string $event_name): bool
+    protected function processMiddleware(Route $route, string $event_name): bool
     {
         $at_least_one_middleware_failed = false;
 
-        foreach ($middleware as $middleware) {
+        $middlewares = $event_name === 'before' ? $route->middleware : array_reverse($route->middleware);
```

Comment:
You could use Dispatcher::FILTER_BEFORE to avoid possible typos

Reply:
Deal